NetKet/Machine/fullconn_layer.hpp

```diff
@@ -0,0 +1,257 @@
+
+ public:
+  using StateType = typename AbstractLayer<T>::StateType;
+  using LookupType = typename AbstractLayer<T>::LookupType;
+
+  /// Constructor
+  FullyConnected(const int input_size, const int output_size,
+                 const bool use_bias = true)
+      : activation_(),
+        usebias_(use_bias),
+        in_size_(input_size),
+        out_size_(output_size) {
+    Init();
+  }
+
+  explicit FullyConnected(const json &pars) : activation_() { Init(pars); }
+
+  void Init(const json &pars) {
+    in_size_ = FieldVal(pars, "Inputs");
+    out_size_ = FieldVal(pars, "Outputs");
+
+    usebias_ = FieldOrDefaultVal(pars, "UseBias", true);
+
+    scalar_bytesize_ = sizeof(std::complex<double>);
+
+    weight_.resize(in_size_, out_size_);
+    bias_.resize(out_size_);
+    din_.resize(in_size_);
+
+    npar_ = in_size_ * out_size_;
+
+    z_.resize(out_size_);
+    a_.resize(out_size_);
+
+    if (usebias_) {
+      npar_ += out_size_;
+    } else {
+      bias_.setZero();
+    }
+    std::string buffer = "";
+
+    InfoMessage(buffer) << "Fully Connected Layer " << in_size_ << " --> "
+                        << out_size_ << std::endl;
+    InfoMessage(buffer) << "# # UseBias = " << usebias_ << std::endl;
+  }
+
+  void Init() {
+    scalar_bytesize_ = sizeof(std::complex<double>);
+
+    weight_.resize(in_size_, out_size_);
+    bias_.resize(out_size_);
+    din_.resize(in_size_);
+
+    npar_ = in_size_ * out_size_;
+
+    z_.resize(out_size_);
+    a_.resize(out_size_);
+
+    if (usebias_) {
+      npar_ += out_size_;
+    } else {
+      bias_.setZero();
+    }
+
+    std::string buffer = "";
+
+    InfoMessage(buffer) << "Fully Connected Layer " << in_size_ << " --> "
+                        << out_size_ << std::endl;
+    InfoMessage(buffer) << "# # UseBias = " << usebias_ << std::endl;
+  }
+
+  void to_json(json &pars) const override {
+    json layerpar;
+    layerpar["Name"] = "FullyConnected";
+    layerpar["UseBias"] = usebias_;
+    layerpar["Inputs"] = in_size_;
+    layerpar["Outputs"] = out_size_;
+    layerpar["Bias"] = bias_;
+    layerpar["Weight"] = weight_;
+
+    pars["Machine"]["Layers"].push_back(layerpar);
+  }
+
+  void from_json(const json &pars) override {
+    if (FieldExists(pars, "Weight")) {
+      weight_ = pars["Weight"];
+    } else {
+      weight_.setZero();
+    }
+    if (FieldExists(pars, "Bias")) {
+      bias_ = pars["Bias"];
+    } else {
+      bias_.setZero();
+    }
+  }
+
+  void InitRandomPars(int seed, double sigma) override {
+    VectorType par(npar_);
+
+    netket::RandomGaussian(par, seed, sigma);
+
+    SetParameters(par, 0);
+  }
+
+  int Npar() const override { return npar_; }
+
+  int Ninput() const override { return in_size_; }
+
+  int Noutput() const override { return out_size_; }
+
+  void GetParameters(VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(pars.data() + k, bias_.data(), out_size_ * scalar_bytesize_);
```

Comment:
Is this `memcpy` for efficiency? This can probably written in a safer way using Eigen methods. Are you sure it is necessary to sacrifice this safety by copying the raw memory?